emake.mk

```diff
@@ -0,0 +1,101 @@
+# Minimum configuration
+#   PACKAGE_BASENAME: the basename that prefixes all Lisp files in this package
+
+ifndef PACKAGE_BASENAME
+$(error PACKAGE_BASENAME is not set)
+endif
+
+# User options
+#   EMACS_ARGS: [extra arguments for each invocation of emacs]
+#   (below; see README)
+PACKAGE_FILE            ?= $(PACKAGE_BASENAME).el
+PACKAGE_LISP            ?= $(filter-out %-pkg.el %-autoloads.el, $(wildcard $(PACKAGE_BASENAME)*.el))
+PACKAGE_TESTS           ?= $(wildcard test/*.el)
+PACKAGE_ARCHIVES        ?= gnu
+PACKAGE_TEST_ARCHIVES   ?= gnu
+
+# Then, make it easy to invoke Emacs with EMake loaded.
+EMAKE = PACKAGE_FILE="$(PACKAGE_FILE)" \
+	PACKAGE_LISP="$(PACKAGE_LISP)" \
+	PACKAGE_TESTS="$(PACKAGE_TESTS)" \
+	PACKAGE_ARCHIVES="$(PACKAGE_ARCHIVES)" \
+	PACKAGE_TEST_DEPS="$(PACKAGE_TEST_DEPS)" \
+	PACKAGE_TEST_ARCHIVES="$(PACKAGE_TEST_ARCHIVES)" \
+	emacs -batch -l emake.el \
+	--eval "(setq enable-dir-local-variables nil)" \
+	$(EMACS_ARGS) \
+	--eval "(emake (pop argv))"
+
+# Set up our phony targets so Make doesn't think there are files by
+# these names.
+.PHONY: clean setup install compile test help
+
+help:                           ## show help
+	@grep -E '(^[A-Za-z_/\.\-]+:.*?##.*$$)|(^##.*$$)' $(MAKEFILE_LIST) \
+		| awk 'BEGIN {FS = ":.*?## "}{printf "\033[32m%-30s\033[0m %s\n", $$1, $$2}' \
+		| sed -e 's/\[32m## */[33m/'
+
+
+# Tell Make how to 'clean' this project
+clean:                          ## clean all generated files
+	rm -f *.elc             # delete compiled files
+	rm -rf .elpa/           # delete dependencies
+	rm -rf .elpa.test/
+clean-all: clean                ## clean all generated files including scripts
+	rm -f emacs-travis.mk
+	rm -f emake.el
+
+## Commands useful for Travis
+
+# Tell Make how to 'setup' this project (e.g., for Travis).  This
+# requires both Emacs to be installed and the `emake.el' script to be
+# available.
+setup: emacs emake.el
+
+install: .elpa/                 ## install dependencies
```

Comment:
I'm not sure about the name of this rule. Usually, `make install` installs the project under a common directory such as /usr/local. What about `install-dependencies`, `setup-dependencies`, `setup-elpa`...?

Reply:
I'm wondering now if a 'target pseudonym' like this is warranted…

Is `make .elpa/` clear?

All the testing targets already require `.elpa/`.

Reply:
to me `make .elpa` is clear enough.